### PR TITLE
Presence absence to FHIR

### PR DESCRIPTION
**Pull request by joverlee521**
Update FHIR ETL to process presence/absence FHIR Bundles.
Major change is loading in an external module with customizations. 
Needs to be deployed with https://github.com/seattleflu/id3c-customizations/pull/142

### CODE REVIEW

**Comment by joverlee521** [comment]
[Re-engineer presence/absence ingest to split bundles from Samplify by sample and insert one FHIR bundle per sample](https://trello.com/c/zfZjRLic/757-re-engineer-presence-absence-ingest-to-split-bundles-from-samplify-by-sample-and-insert-one-fhir-bundle-per-sample)

File: lib/id3c/cli/command/etl/fhir.py [hunk 1]
```diff
@@ -940,6 +924,34 @@ def observation_value(observation: Observation) -> bool:
             details = details)
 
 
+def observation_value(observation: Observation) -> bool:
```
Comment (kairstenfay) [inline lib/id3c/cli/command/etl/fhir.py:927]: Nitpick but technically this would return `Optional[bool]`

File: lib/id3c/cli/command/etl/fhir.py [hunk 2]
```diff
@@ -965,3 +1009,22 @@ def mark_processed(db, fhir_id: int, entry = {}) -> None:
...
+
+# Load all extra customizations for the INTERNAL_SYSTEM.
+# Needs to be at the end to avoid dependency errors.
+extensions = list(pkg_resources.iter_entry_points("id3c.cli.command.etl.fhir", INTERNAL_SYSTEM))
+
+# Only expects one module for the INTERNAL_SYSTEM.
+assert len(extensions) <= 1, \
+    f"The FHIR ETL only expects one extension from the internal system «{INTERNAL_SYSTEM}»"
+
+extension = extensions[0]
```
Comment (kairstenfay) [inline lib/id3c/cli/command/etl/fhir.py:1022]: I think we'll want to specially handle an empty extensions list.

**Review by kairstenfay — COMMENTED** [review]
Thanks for submitting this ages ago, Jover! Sorry it took so long for me to get around to reviewing it. I am going to look at the id3c-customizations changes next, but figured I would submit some interim feedback now.

File: lib/id3c/cli/command/etl/fhir.py (hunk 2 above)
Comment (kairstenfay) [reply lib/id3c/cli/command/etl/fhir.py:1022]: When I run this locally on old presence-absence results, I get
```py
$ pipenv run id3c etl presence-absence
[2020-11-18 11:48:40] Uncaught exception:
Traceback (most recent call last):
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/bin/id3c", line 11, in <module>
    load_entry_point('id3c', 'console_scripts', 'id3c')()
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pkg_resources/__init__.py", line 490, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2853, in load_entry_point
    return ep.load()
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2453, in load
    return self.resolve()
  File "/home/kairsten/.local/share/virtualenvs/id3c-customizations-OmmbKeV1/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2459, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/kairsten/seattleflu/id3c/lib/id3c/cli/__init__.py", line 21, in <module>
    from .command import *
  File "/home/kairsten/seattleflu/id3c/lib/id3c/cli/command/etl/__init__.py", line 484, in <module>
    from . import *
  File "/home/kairsten/seattleflu/id3c/lib/id3c/cli/command/etl/fhir.py", line 1022, in <module>
    extension = extensions[0]
IndexError: list index out of range
```

File: lib/id3c/cli/command/etl/fhir.py [hunk 3]
```diff
@@ -226,6 +227,11 @@ def process_diagnostic_report_bundle_entry(db: DatabaseSession, bundle: Bundle,
         if not is_collection_identifier and not sample:
             raise SampleNotFoundError("No sample with identifier «{specimen_identifier.uuid}» found.")
 
+        if specimen:
+            # If there is a Specimen Resource, check for potential updates
+            # that need to be handled by FHIR ETL customizations.
+            sample = INTERNAL_CUSTOMIZATIONS.check_specimen_for_update(db, sample, specimen)
+
```
[the remark this answers is not in the document]
Comment (joverlee521) [reply lib/id3c/cli/command/etl/fhir.py:234]: In the DiagnosticReport Resource, `resource.specimen` is a list of specimen references. Our internal FHIR bundles usually does have a specimen resource in the same bundle, but external bundles like the ones from Ellume only have a specimen reference and no specimen resource.

File: lib/id3c/cli/command/etl/fhir.py [hunk 4]
```diff
@@ -223,6 +227,11 @@ def process_diagnostic_report_bundle_entry(db: DatabaseSession, bundle: Bundle,
         if not is_collection_identifier and not sample:
             raise SampleNotFoundError("No sample with identifier «{specimen_identifier.uuid}» found.")
 
+        if specimen:
+            # If there is a Specimen Resource, check for potential updates
+            # that need to be handled by FHIR ETL customizations.
+            sample = INTERNAL_CUSTOMIZATIONS.check_specimen_for_update(db, sample, specimen)
```
Comment (joverlee521) [inline lib/id3c/cli/command/etl/fhir.py:233]: Now that I'm looking at this again, it feels weird that I'm calling a custom function within core. Is there another way of doing this?

**Comment by joverlee521** [comment]
Closing in favor of simpler unbundling method

**Pull request closed without merging**